src/solvers/dg_p4est/containers_2d.jl

```diff
@@ -61,37 +69,82 @@ function calc_node_coordinates!(node_coordinates,
 end
 
 
-# Iterate over all interfaces and extract inner interface data to interface container
-# This function will be passed to p4est in init_interfaces! below
-function init_interfaces_iter_face(info, user_data)
-  if info.sides.elem_count != 2
-    # Not an inner interface
-    return nothing
-  end
-
-  sides = (unsafe_load_sc(p4est_iter_face_side_t, info.sides, 1),
-           unsafe_load_sc(p4est_iter_face_side_t, info.sides, 2))
+# A helper struct used in initialization methods below
+mutable struct InitSurfacesIterFaceUserData{Interfaces, Mortars, Boundaries, Mesh}
+  interfaces  ::Interfaces
+  interface_id::Int
+  mortars     ::Mortars
+  mortar_id   ::Int
+  boundaries  ::Boundaries
+  boundary_id ::Int
+  mesh        ::Mesh
+end
 
-  if sides[1].is_hanging == true || sides[2].is_hanging == true
-    # Mortar, no normal interface
-    return nothing
-  end
+function InitSurfacesIterFaceUserData(interfaces, mortars, boundaries, mesh)
+  return InitSurfacesIterFaceUserData{
+    typeof(interfaces), typeof(mortars), typeof(boundaries), typeof(mesh)}(
+      interfaces, 1, mortars, 1, boundaries, 1, mesh)
+end
 
-  # Unpack user_data = [interfaces, interface_id, mesh] and increment interface_id
-  ptr = Ptr{Any}(user_data)
-  data_array = unsafe_wrap(Array, ptr, 3)
-  interfaces = data_array[1]
-  interface_id = data_array[2]
-  data_array[2] = interface_id + 1
-  mesh = data_array[3]
+function init_surfaces_iter_face(info, user_data)
+  # Unpack user_data
+  data = unsafe_pointer_to_objref(Ptr{InitSurfacesIterFaceUserData}(user_data))
```

Comment:
Citing from my blog post:
> Providing generic Julia code as callback via C can be tricky and we still have some type instabilities left for the future. We could fix them by passing Julia closures to C, but that’s not supported on all platforms [such as ARM](https://docs.julialang.org/en/v1/manual/calling-c-and-fortran-code/#Closure-cfunctions). Since there are ARM clusters, using this feature is not an option for us in Trixi.jl.

Sadly, I don't really see an easy and simple way to make this part type-stable since `InitSurfacesIterFaceUserData` has quite some type parameters and we use several different versions of it. I don't see a better way to obtain type stability than to use some code generation. Using closures is out of questions as mentioned above. Hence, I guess we would need to use more dirty tricks and I don't know whether that's worth the effort right now. It might be possible to define a function knowing everything about the types and eval that on the fly, but that seems rather fishy to me...

Reply:
The basic problem is that we can't use Julia's strong type system and multiple dispatch via the detour in C...